utility/include/isobus/utility/thread_synchronization.hpp

```diff
@@ -88,16 +98,368 @@ class LockFreeQueue
+	class RecursiveMutex : public Mutex
+	{
+	protected:
+		/// @brief Checks if the mutex is ready to be used.
+		/// Initializes the mutex if it's not.
+		/// @returns true if the mutex is ready to be used, false otherwise.
+		bool ready() override
+		{
+			if (nullptr == handle)
+			{
+				const osMutexAttr_t attributes = {
+					nullptr,
+#ifdef osCMSIS_FreeRTOS // FreeRTOS doesn't support robust mutexes
+					osMutexPrioInherit | osMutexRecursive,
+#else
+					osMutexPrioInherit | osMutexRobust | osMutexRecursive,
+#endif
+					nullptr,
+					0
+				};
+				handle = osMutexNew(&attributes);
+			}
+			return nullptr != handle;
+		}
+	};
+
+	template<class T>
+	/// @brief A class to automatically lock and unlock a mutex when the scope ends.
+	/// Meant for systems with no support for std::lock_guard.
+	class LockGuard
+	{
+	public:
+		/// @brief Constructor for the LockGuard class.
+		/// @param mutex The mutex to lock.
+		/// @details Locks the mutex when the scope starts.
+		/// Unlocks the mutex when the scope ends.
+		LockGuard(T *mutex) :
+		  lockable(mutex)
+		{
+			lockable->lock();
+		}
+
+		/// @brief Destructor for the LockGuard class.
+		/// @details Unlocks the mutex when the scope ends.
+		~LockGuard()
+		{
+			lockable->unlock();
+		}
+
+	private:
+		T *lockable; ///< The mutex to lock and unlock.
+	};
+
+	/// @brief A class that emulates a condition variable using CMSIS thread flags
+	class ConditionVariable
+	{
+	public:
+		ConditionVariable() = default;
+
+		void wait_for(Mutex &, std::uint32_t timeout)
+		{
+			targetThread = osThreadGetId();
+			osThreadFlagsWait(0x00000001, osFlagsWaitAny, timeout);
+		}
+
+		void notify_all()
+		{
+			osThreadFlagsSet(targetThread, 0x00000001);
+		}
+
+	private:
+		osThreadId_t targetThread = NULL;
+	};
+
+	class Thread
+	{
+	public:
+		Thread(osThreadFunc_t threadFunction, void *argument)
+		{
+			osThreadAttr_t attributes = {
+				NULL,
+				osThreadDetached,
+				NULL,
+				0,
+				NULL,
+				0,
+				osPriorityNormal,
+				0,
+				0
+			};
+			myID = osThreadNew(threadFunction, argument, &attributes);
+			if (NULL == myID)
+			{
+				while (true)
+				{
+					// If your code is stuck in here, it means you don't have enough OS memory left to spawn the thread.
+					// Consider increasing it.
+				}
+			}
+			else
+			{
+				isJoinable = true;
+			}
+		}
+
+		bool joinable() const
+		{
+			return isJoinable;
+		}
+
+		void join()
+		{
+			if (isJoinable)
+			{
+				osThreadTerminate(myID);
+			}
+		}
+
+	private:
+		osThreadId_t myID = NULL;
+		bool isJoinable = false;
+	};
+} // namespace isobus
+
+namespace std
+{
+	using mutex = isobus::Mutex;
+	using recursive_mutex = isobus::RecursiveMutex;
+	using condition_variable = isobus::ConditionVariable;
+} // namespace std
+
+#define LOCK_GUARD(type, x) const LockGuard<type> x##Lock(&x)
+
+/// @brief A template class for a lock free queue.
+/// @tparam T The item type for the queue.
+template<typename T>
+class LockFreeQueue
```

Comment:
nit: LockFreeQueue is already in the file below. I think it makes sense if we just use a separate `#ifdef` directive there that covers both cases